libraries/AP_IOMCU/iofirmware/iofirmware.cpp

```diff
@@ -147,22 +207,43 @@ static void dma_tx_end_cb(hal_uart_driver *uart)
 /* replacement for ChibiOS uart_lld_serve_interrupt() */
 static void idle_rx_handler(hal_uart_driver *uart)
 {
+#if defined(STM32H7)
+    #define USART_SR_REG uart->usart->ISR
+    #define USART_DR_REG uart->usart->RDR
+    #define USART_SR_LBD USART_ISR_LBDF
+    #define USART_SR_ORE USART_ISR_ORE
+    #define USART_SR_NE USART_ISR_NE
+    #define USART_SR_FE USART_ISR_FE
+    #define USART_SR_PE USART_ISR_PE
+    #define USART_SR_TC USART_ISR_TC
+    #define USART_SR_IDLE USART_ISR_IDLE
+    #define CLEAR_ISR_FLAG(flag) uart->usart->ICR = USART_ICR_##flag##CF
+    #define SEND_BREAK() uart->usart->RQR = USART_RQR_SBKRQ
+    #define CLEAR_IDLE() uart->usart->ICR = USART_ICR_IDLECF
+    #define CLEAR_ERRORS() uart->usart->ICR = USART_ICR_ORECF | USART_ICR_NECF | USART_ICR_FECF | USART_ICR_PECF
+#else
+    #define USART_SR_REG uart->usart->SR
+    #define USART_DR_REG uart->usart->DR
+    #define CLEAR_ISR_FLAG(flag) uart->usart->SR = ~USART_SR_##flag
+    #define SEND_BREAK() uart->usart->CR1 |= USART_CR1_SBK
+    #define CLEAR_IDLE() (void)uart->usart->DR
+    #define CLEAR_ERRORS() (void)USART_DR_REG
+#endif
+
     volatile uint16_t sr;
-    sr = uart->usart->SR; /* SR reset step 1.*/
+    sr = USART_SR_REG; /* SR reset step 1.*/
     uint32_t cr1 = uart->usart->CR1;
 
     if (sr & (USART_SR_LBD | USART_SR_ORE |	/* overrun error - packet was too big for DMA or DMA was too slow */
               USART_SR_NE |		/* noise error - we have lost a byte due to noise */
               USART_SR_FE |
               USART_SR_PE)) {		/* framing error - start/stop bit lost or line break */
 
-        (void)uart->usart->DR;  /* SR reset step 2 - clear ORE | FE.*/
-
+        CLEAR_ERRORS(); /* SR reset step 2 - clear ORE | FE.*/
         /* send a line break - this will abort transmission/reception on the other end */
         chSysLockFromISR();
-        uart->usart->SR = ~USART_SR_LBD;
-        uart->usart->CR1 = cr1 | USART_CR1_SBK;
-
+        CLEAR_ISR_FLAG(LBD);
+        SEND_BREAK();
```

Comment:
this is a slight change in behaviour because this was a local variable for threading reasons. Ideally would preserve the current semantics

Reply:
I am going to change it back to original as it doesn't really matter in this context.

But the change is actually correct. Firstly a thread cannot interrupt an interrupt, another higher priority interrupt can. Secondly, only bit we want to set in CR1 is Line Break, rest of the bits we don't want to touch. With the older method there's a chance we can end up doing just that.
Basically if there was another interrupt or peripheral event (between where we store cr1 and call SEND_BREAK) where the CR1 is changed we will end up changing it again unknowingly. Fortuntely we don't mess with CR1 outside of init and this interrupt, so it is a non-issue. But in general this can cause issue.